ufl/cell.py

```diff
@@ -1,298 +1,456 @@
 #
 # This file is part of UFL (https://www.fenicsproject.org)
 #
 # SPDX-License-Identifier:    LGPL-3.0-or-later
 
+from __future__ import annotations
 import functools
 import numbers
+import typing
 
-import ufl.cell
-from ufl.core.ufl_type import attach_operators_from_hash_data
+from ufl.core.ufl_type import UFLObject
+from abc import abstractmethod
+
+try:
+    from typing import Self
+except ImportError:
+    # Hack to get this to work pre Python 3.11
+    # Delete this after 04 Oct 2026 (Python 3.10 end of life)
+    # After this date, can also replace "self: Self," with "self,
+    from typing import TypeVar
+    Self = TypeVar("Self", bound="AbstractCell")
 
-# Export list for ufl.classes
 __all_classes__ = ["AbstractCell", "Cell", "TensorProductCell"]
 
 
-# --- The most abstract cell class, base class for other cell types
+class AbstractCell(UFLObject):
+    """A base class for all cells."""
+    @abstractmethod
+    def topological_dimension(self) -> int:
+        """Return the dimension of the topology of this cell."""
+
+    @abstractmethod
+    def geometric_dimension(self) -> int:
+        """Return the dimension of the geometry of this cell."""
+
+    @abstractmethod
+    def is_simplex(self) -> bool:
+        """Return True if this is a simplex cell."""
+
+    @abstractmethod
+    def has_simplex_facets(self) -> bool:
+        """Return True if all the facets of this cell are simplex cells."""
+
+    @abstractmethod
+    def _lt(self: Self, other: Self) -> bool:
+        """Define an arbitrarily chosen but fixed sort order for all instances of this type with the same dimensions."""
+
+    @abstractmethod
+    def num_sub_entities(self, dim: int) -> int:
+        """Get the number of sub-entities of the given dimension."""
+
+    @abstractmethod
+    def sub_entities(self, dim: int) -> typing.List[AbstractCell]:
+        """Get the sub-entities of the given dimension."""
+
+    @abstractmethod
+    def sub_entity_types(self, dim: int) -> typing.List[AbstractCell]:
+        """Get the unique sub-entity types of the given dimension."""
+
+    @abstractmethod
+    def cellname(self) -> str:
+        """Return the cellname of the cell."""
+
+    @abstractmethod
+    def reconstruct(self, **kwargs: typing.Any) -> Cell:
+        """Reconstruct this cell, overwriting properties by those in kwargs."""
+
+    def __lt__(self, other: AbstractCell) -> bool:
+        """Define an arbitrarily chosen but fixed sort order for all cells."""
+        if type(self) == type(other):
+            s = (self.geometric_dimension(), self.topological_dimension())
+            o = (other.geometric_dimension(), other.topological_dimension())
+            if s != o:
+                return s < o
+            return self._lt(other)
+        else:
+            if type(self).__name__ == type(other).__name__:
+                raise ValueError("Cannot order cell types with the same name")
+            return type(self).__name__ < type(other).__name__
```

Comment:
Could one define a total order over cells given information about the cell complex they represent? Or is that insufficient because you might have two different cell implementations that both represent a hexahedron (say).

Reply:
> you might have two different cell implementations that both represent a hexahedron (say).

Exactly that

Reply:
Is it sufficient (for sorting) to only define `__lt__` or does one need to provide `__eq__` and then `@functools.total_ordering` ?

Reply:
Good question. `__eq__` is defined as part of `UFLObject` (formerly `@attach_operators_from_hash_data`). Only this and `__lt__` were previously defined so I guess it's enough for what UFL currently does, but adding `@functools.total_ordering` probably isn't a bad idea